lib/SILOptimizer/Mandatory/MoveOnlyChecker.cpp

```diff
@@ -110,10 +112,66 @@ void MoveOnlyChecker::checkObjects() {
     return;
   }
 
+  completeObjectLifetimes(moveIntroducersToProcess.getArrayRef());
+
   MoveOnlyObjectChecker checker{diagnosticEmitter, domTree, poa, allocator};
   madeChange |= checker.check(moveIntroducersToProcess);
 }
 
+void MoveOnlyChecker::completeObjectLifetimes(
+    ArrayRef<MarkUnresolvedNonCopyableValueInst *> insts) {
+// TODO: Delete once OSSALifetimeCompletion is run as part of SILGenCleanup.
+  OSSALifetimeCompletion completion(fn, domTree);
+
+  // Collect all values derived from each mark_unresolved_non_copyable_value
+  // instruction via ownership instructions and phis.
+  ValueWorklist transitiveValues(fn);
+  for (auto *inst : insts) {
+    transitiveValues.push(inst);
+  }
+  while (auto value = transitiveValues.pop()) {
+    for (auto *use : value->getUses()) {
+      auto *user = use->getUser();
+      switch (user->getKind()) {
+      case SILInstructionKind::BeginBorrowInst:
+      case SILInstructionKind::CopyValueInst:
+      case SILInstructionKind::MoveValueInst:
+        transitiveValues.pushIfNotVisited(cast<SingleValueInstruction>(user));
+        break;
+      case SILInstructionKind::BranchInst: {
+        PhiOperand po(use);
+        transitiveValues.pushIfNotVisited(po.getValue());
+        break;
+      }
```

Comment:
do we need forwarding operations handling here ?

Reply:
Great point, seems like that would be needed at least for forwards of borrows.  I guess forwards of owned values, at least those that preserve ownership, wouldn't need to be considered.

Reply:
Picked up 2bb32129955b8f392ee42deb2725195f34b3e9e1 with your `visitForwardedValues` API in this PR and called it here.